### DISCUSSION TITLE
GenericType syntax

### DISCUSSION THREAD

**Discussion by xialvjun** [koka-lang/koka#332] [Q&A] [answered]
It seems the doc didn't say it clearly.

```koka
fun fold(t : tree<k,a>, acc : b, f : (k, a, b) -> b) : b
  match t
    Node(_,l,k,v,r) -> r.fold( f(k,v,l.fold(acc,f)), f)
    Leaf            -> acc
```

Yeah, I know `k, a, b` all are generic type.  But 

1. how koka know `acc : b` is a generic type rather than there is a type named `b` ? If `b` is a generic type, will `fold` function be changed when I write `type b = xxx` above it ?
2. how koka know `f : (k, a, b) -> b`, the `k, a, b` are all generic type, what if I define `f` before `t : tree<k,a>, acc : b`
3. what's the standard or officially recommended way to write generic type ? `fun fold<k, a, b>(t: tree<k, a>, acc: b, f: (k, a, b) -> b): b` or just `fun fold(t : tree<k,a>, acc : b, f : (k, a, b) -> b) : b`

Or generic type and normal type share the same namespace? All undefined name used in type position are generic type?

**Comment by JanBeh** [comment, marked as answer]
I'm unsure, but I think the book explains it in ["2.2. Effect typing"](https://koka-lang.github.io/koka/doc/book.html#why-effects):

> Single letter types are polymorphic (aka, generic), and Koka infers that you `map` from a list of elements `a` to a list of elements of type `b`.

So the fact that the type name consists of a *single letter* is relevant here.

The same is explained in ["3.2.3. Polymorphic effects"](https://koka-lang.github.io/koka/doc/book.html#sec-polymorphic-effects):

> Many functions are polymorphic in their effect. For example, the [`map`](https://koka-lang.github.io/koka/doc/std_core.html#map) function applies a function `f` to each element of a (finite) list. As such, the effect depends on the effect of `f`, and the type of `map` becomes:
> 
> ```
> map : (xs : list<a>, f : (a) -> e b) -> e list<b>
> ```
> 
> We use single letters (possibly followed by digits) for polymorphic types. Here, the `map` functions takes a list with elements of some type `a`, and a function `f` that takes an element of type `a` and returns a new element of type `b`. The final result is a list with elements of type `b`. Moreover, the effect of the applied function `e` is also the effect of the `map` function itself; indeed, this function has no other effect by itself since it does not diverge, nor raises exceptions.

I think the documentation could be a bit more clear that it's not just a convention but a syntactical difference:

```
fun poly1( arg : t ) : t
  arg

fun poly2<ty>( arg : ty ) : ty
  arg
```

You don't need the `<…>` quantifier for the `poly1` function above because `t` is a *single letter* name (optionally followed by digits). If the type consists of more than a single letter as in example `poly2`, then you need the explicit quantifier `<ty>`.

**Reply by xialvjun** [reply]
Thanks. So `single letters (possibly followed by digits)` or `complete syntax <...>`.
